pkg/utils/netfilter.go

```diff
@@ -0,0 +1,45 @@
+// Copyright 2017 CNI authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package utils
+
+import (
+	"github.com/coreos/go-iptables/iptables"
+	"github.com/danwinship/nftables"
+)
+
+// SupportsIPTables tests whether the system supports using netfilter via the iptables API
+// (whether via "iptables-legacy" or "iptables-nft"). (Note that this returns true if it
+// is *possible* to use iptables; it does not test whether any other components on the
+// system are *actually* using iptables.)
+func SupportsIPTables() bool {
+	ipt, err := iptables.NewWithProtocol(iptables.ProtocolIPv4)
```

Comment:
is going to be important the protocol?
as if  `SupportsIP6Tables`?

Reply:
(So first off, we only have to worry about "`iptables` works but `ip6tables` doesn't". IMO there's no plausible scenario where the reverse is true.)

In kube-proxy we have the problem that we don't really know for sure at startup if we're going to be dealing with IPv4, IPv6, or dual-stack Services. So it's plausible that we'll try to use `ip6tables` and then discover it doesn't work, because the administrator hadn't expected us to do anything with IPv6.

The CNI plugins don't need to worry about this: they don't have "startup time"; they only run when there's a specific pod IP that they need to work with. So if `iptables` is available and `ip6tables` isn't, then either (a) it doesn't matter because we'll never see an IPv6 IP anyway, or (b) the host is broken and we _should_ return errors.

So I think it's fine as-is?

Reply:
fair